src/models/orders.rs

```diff
@@ -13,8 +13,8 @@ use std::cmp::PartialOrd;
 pub struct Order {
     pub sell_amount: U256,
     pub buy_amount: U256,
-    pub current_sell_amount: U256,
-    pub current_buy_amount: U256,
+    pub current_sell_amount: Option<U256>,
```

Comment:
Why is this optional now? Also is this remaining sell amount, or amount filled so far (maybe we can rename for clarification)?

Reply:
It is intended to be the sell_amount reduce by min(allowance(), balanceOf()). We need to store this information somewhere.

It is optional so that in the warp API call, we can leave these values out. They will only later be processed by our allowance and balanceOf readers.

Reply:
> It is optional so that in the warp API call, we can leave these values out

I personally think that this is a sign that we shouldn't use the exact same Order model for posting orders and for processing them (we are shoehorning one model to work in two places).

We could keep the raw Order struct for posting (without `current_sell_amount`) and have a MatchableOrder for solving, which consists of the raw Order and an extra field (available_sell_balance), which can be the min of balanceOf and allowance.

I understand why we need to look at the "available sell balance". However, why do we need to look at current_buy_amount?

Reply:
> I personally think that this is a sign that we shouldn't use the exact same Order model for posting orders and for processing them (we are shoehorning one model to work in two places).

This is how I had it implemented before. I have also seen it like that in dex-services. But it introduces redundant code and types. It feels strange that if I am modifying the order model to introduce a new field, that I need to do it on several structs. Isn't it much easier to work in Options, in order to reduce the redundancy?

What is the advantage of introducing two order models?

>However, why do we need to look at current_buy_amount?
Yeah, current_buy_amount can always be calculated from current_sell_amount, sell_amount and buy_amount. I thought it would be easier to store this value than recalculating it. But, now that you are pointing it out, I will change it.

Though, most likely, we should also remove current_sell_amount, as a sell balance not easily be assigned directly to one order. If one user has two orders selling the same asset and the available balance is less then the sum of the sell_amounts, we need to deal with that. But I will do it in a later PR.

Reply:
> What is the advantage of introducing two order models?

The advantage is normalisation and scoping a model to each use case. It can become very hard to understand a data model where a lot of fields are optional, but implicitly they have a semantic (e.g. if current_sell_amount is set so is current_buy_amount). SAP's data model is a great example (if you have ever worked with it), where tables have up to 1000 columns (most of them nullable) and it requires expert knowledge to know which columns go with what others in what state of the system.

Such invariants can be enforced at compile time by using tightly scoped models. Converting between them is trivial code (which might seem redundant but has very little chance of error). Keeping track of different field semantic in different parts of the codebase 6 months down the line is much more prone to error.

I'm also not saying we should copy all fields, I think we can use composition where

```rust
struct MatchableOrder {
  raw_order: Order,
  available_sell_amount: U256
}
```

Reply:
I agree with @fleupold, The smaller the scope of a type, the better. If we use `Option<U256>` then we might need to introduce `None` handling code in places where it is impossible to be `None`, or introduce `unwrap`s which are brittle to refactoring. By encoding the exact "shape" of the data in its type, and using different types where the data has different "shapes" is a little nicer.

Note that `serde` has a `#[serde(flatten)]` feature, so the JSON could even look the same!